### PR TITLE
Date picker in RPA.Dialogs

### LINKED ISSUE #219: RPA.Dialogs date picker
Add the possibility of picking up a date from the dialog UI

- [x] Support internally a date picking React element (in **robocorp-dialog** dependency)
- [ ] `Add Date Input` keyword in RF displaying and collecting data from the element

### Limitations
- Can't easily pick a date far away from the current year (or provided default)
- Can't input date as text directly in the entry (read-only value property)

### PR DESCRIPTION
**Pull request by cmin764**
Closes #219 

- Adds support for date picking widget under `RPA.Dialogs`
- Robot and unit-tests
- See [limitations](https://github.com/robocorp/rpaframework/issues/219)

Tested with this:
```robotframework
Get Birthdate
    Add icon      Success
    Add Date Input    birthdate    26/04/1993
    ${result} =    Run dialog    title=Success
    Log To Console    ${result.birthdate}
```

### CODE REVIEW

File: packages/dialogs/src/RPA/Dialogs/library.py [hunk 1]
```diff
@@ -696,6 +697,53 @@ def add_drop_down(
...
+        if default:
+            try:
+                datetime.strptime(default, format)
+            except Exception as exc:
+                raise ValueError(f"Invalid default date with value {default!r}") from exc
+        else:
+            default = datetime.utcnow().strftime(format)
+
+        element = {
+            "type": "input-datepicker",
```
Comment (cmin764) [inline packages/dialogs/src/RPA/Dialogs/library.py:739]: We're blocked here by this other robocorp-dialogs [PR](https://github.com/robocorp/robocorp-dialog/pull/1).

File: packages/dialogs/src/RPA/Dialogs/library.py [hunk 2]
```diff
@@ -696,6 +697,53 @@ def add_drop_down(
 
         self.add_element(element)
 
+    @keyword("Add date-picker", tags=["input"])
```
Comment (mikahanninen) [inline packages/dialogs/src/RPA/Dialogs/library.py:700]: I think the name for the  keyword should actually be **Add Date Input** to follow naming convention within current library implementation.

**Review by mikahanninen — COMMENTED** [review]
It would be good to show at least as "todo" comments in the code that the keyword still requires support for "value", "min" and "max" for the date input field.

**Comment by cmin764** [comment]
> show at least as "todo" comments in the code that the keyword still requires support for "value", "min" and "max" for the date input field.

Agree, but I feel this wouldn't make sense anymore if we use the WIP React [datepicker](https://robocorp-design-system-jq08uln2s-robocorp-dev.vercel.app/component/datepicker) (usage in [PR](https://github.com/robocorp/robocorp-dialog/pull/1)), right? As of the following:
- `value` is handled internally (picks up `default`'s value if provided)
- `min` & `max` not supported yet by the design system

----

Still, I can add some TODOs if we plan to add extra support in the future, but not sure yet what to write :).

File: packages/dialogs/src/RPA/Dialogs/library.py (hunk 1 above)
Comment (cmin764) [reply packages/dialogs/src/RPA/Dialogs/library.py:742]: Not blocked anymore.

**Comment by cmin764** [comment]
_Access rights required_

After merge, I should be enabled to publish the new `0.4.0` version to PyPI using my [user](https://pypi.org/user/cmiN/).

```console
UploadError

HTTP Error 403: The user 'cmiN' isn't allowed to upload to project 'rpaframework-dialogs'. See https://pypi.org/help/#project-name for more information.
```

File: packages/dialogs/src/RPA/Dialogs/library.py [hunk 3]
```diff
@@ -696,6 +697,59 @@ def add_drop_down(
 
         self.add_element(element)
 
+    @keyword("Add Date Input", tags=["input"])
+    def add_date_input(
+        self,
+        name: str,
+        default: Optional[str] = None,
+        label: Optional[str] = None,
```
Comment (osrjv) [inline packages/dialogs/src/RPA/Dialogs/library.py:705]: I would just drop this from the public API if it isn't supported now. Optional arguments can always be added later.

**Comment by osrjv** [comment]
Seems to work quite nice!

There's something here that I didn't think of earlier. The current result is a string in the DD/MM/YYYY format, right? What would be the way for a user to actually use that as a date in their robot? Should we post-process that somewhere into a python datetime object?

**Comment by cmin764** [comment]
> Seems to work quite nice!
> 
> There's something here that I didn't think of earlier. The current result is a string in the DD/MM/YYYY format, right? What would be the way for a user to actually use that as a date in their robot? Should we post-process that somewhere into a python datetime object?

Yes, the date is returned as a string and I guess the dev could use some date [conversion](https://robocorp.com/docs/libraries/built-in/datetime/keywords#convert-date) mechanism to get an object out of it.
```robotframework
Library           DateTime
...
Convert date test
    ${date_obj} =    Convert Date    26/04/1993    date_format=%d/%m/%Y    result_format=datetime
    Log To Console    ${date_obj}

----
> Get Birthdate                                                         1993-04-26 00:00:00
```
I believe devs got used to expect strings out of the inputs, but if you feel is better to return the object directly, we can do it. If "yes", any idea on how to convert this at its best? I believe we can add a check for the "input-datepicker" element under _packages/dialogs/src/RPA/Dialogs/dialog.py_ so we can convert it into an object from a string when such element is occured.

**Comment by osrjv** [comment]
I guess strings make more sense when we implement the `format` argument, so they can match it with what the system is expecting. There are some exceptions in the results already, like checkboxes being actual booleans.

The parsing would probably be in `src/RPA/Dialogs/dialog.py:143` where the results are parsed and compared to the given elements. However, maybe it can be left to the user. An example in the keyword would be nice though, e.g. how to split year/mont/day from the result.

**Comment by cmin764** [comment]
> I guess strings make more sense when we implement the `format` argument, so they can match it with what the system is expecting. There are some exceptions in the results already, like checkboxes being actual booleans.
> 
> The parsing would probably be in `src/RPA/Dialogs/dialog.py:143` where the results are parsed and compared to the given elements. However, maybe it can be left to the user. An example in the keyword would be nice though, e.g. how to split year/mont/day from the result.

Now the keyword returns a `date` object and accepts both `date[time]` and `str` as `default`. Tested manually too with:
```robotframework
Get Birthdate docstring
    Add heading       Enter your birthdate
    Add Date Input    birthdate    default=26/04/1993
    ${result} =       Run dialog
    Log To Console    User birthdate year should be: ${result.birthdate.year}
```

**Pull request merged**